[2단계 - 출석 구현] 머랭(김대현) 미션 제출합니다.

cookie-meringue opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
step1 에서 작성했고, step2 에서는 따로 작성하지 않았습니다!

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
step1 에서 객체지향적으로 리팩터링하는 과정에서 리소스가 많이 들었습니다.
그 과정에서, 제가 지금 집중해야 하는 것이 OOP 인지 TDD 인지에 대한 생각이 들었습니다.
출석 미션은 TDD 를 위한 미션이므로, 지금은 TDD 식으로 개발하는 것을 학습하는 것이 가장 우선적이 목표라는 생각을 했습니다.
이 생각으로 인해 우선 프로그램의 동작에 초점을 두었고, TDD 사이클을 지켜나가는 데에 주된 리소스를 사용했습니다.

따라서, 이번 step2 코드는 객체지향적이지 못한 부분이 많습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

### 1. 객체지향적인 부분은 최소한으로 봐주셨으면 합니다!
- 제 코드는 제가 봐도 만족스럽지 못합니다. 위에 언급되어 있지만, 저는 step2 에서 객체지향보다는 TDD 에 초점을 두었습니다.
- TDD 적인 리뷰를 먼저 해 주시면 월요일까지 차근차근 객체지향적으로 리팩터링해나가겠습니다!

### 2. TDD 적인가?
- TDD 개발 사이클을 지키기 위해 우선 각 객체에 대한 설계를 가장 먼저 했습니다.
- 제가 생각한 개발 사이클은 다음과 같았습니다.
- 1. 기능 요구사항을 통해 객체 설계하기
- 2. 레드 테스트 작성하기
- 3. 레드 테스트가 통과하도록 구현하기(그린)
- 4. 리팩터링
- 이 과정을 사용해 개발하자고 마음먹었을 때 까지만 해도 TDD 가 나쁘지 않은 개발 방식이라는 생각이 들었습니다. 이론적으로는 크게 문제가 없었기 때문입니다.
- 그러나, 클래스를 설계하는 과정에서 너무 세세하게 설계하지 않았나? 하는 생각이 들었습니다.
- 클래스를 너무 세세하게 설계한 결과, 오히려 한 기능을 구현하기 위해 만들어야 하는 클래스가 많아졌고, 결과적으로 구현 시간이 너무 오래 걸렸습니다.
- 또한, 이미 작성해놓은 테스트가 많으니 구조적 리팩터링을 적용하려면 리소스가 두 배로 들었습니다(프로덕션 코드 리팩터링과 더불어 테스트 코드도 리팩터링해야 했기 때문입니다)
- 앤지는 클래스를 어느 수준까지 설계하는지 궁금합니다.

### 3. TDD 로 개발하니 오히려 개발 시간이 너무 길었습니다.
- 저는 일반적으로, 프로덕션 코드를 직접 작성해보며 전체적인 틀을 구상해나갑니다.
- 이런저런 시도를 해 보며 어떤 게 더 객체지향적일지 생각합니다.
- 하지만, 테스트를 먼저 작성해야 하는 TDD 방식이 오히려 걸림돌이 되는 것 같았습니다. 설계를 하나 적용해보려고 하면 테스트부터 작성해야 하니 손이 잘 가지 않았고, 개발이 즐겁지 않게 느껴졌습니다.
- 이 과정에서 저는 두 가지 추론을 했습니다.
- 1. 내가 테스트 코드를 너무 잘 짜려고 해서 테스트 코드에 쏟는 리소스가 큰 것은 아닌가?
- 2. 내가 커밋을 너무 분리하려고 해서 각 커밋을 찍는 리소스가 큰 것은 아닌가?
- 제 추론 중, 해당되는 것이 있다면 알려주시면 감사하겠습니다.

cookie-meringue committed:
docs: 기능 요구사항 작성

cookie-meringue committed:
docs: Model 구현 클래스 목록 작성

cookie-meringue committed:
chore: Application 생성

- 쓸모없는 .gitkeep 제거.

cookie-meringue committed:
docs: 크루 이름 정책 구현사항 추가

cookie-meringue committed:
feat: 크루 클래스 작성

cookie-meringue committed:
test(red): 크루 이름 검증 레드 테스트 작성

cookie-meringue committed:
feat(green): 크루 닉네임 검증 기능 구현

cookie-meringue committed:
docs: 구현 우선순위 변경

cookie-meringue committed:
docs: Campus 구현 상세 수정

cookie-meringue committed:
feat: 캠퍼스 운영 정책 클래스 작성

cookie-meringue committed:
test(red): 캠퍼스 운영 시간 판별 레드 테스트 작성

cookie-meringue committed:
feat(green): 캠퍼스 운영 시간 판별 기능 구현

cookie-meringue committed:
feat: 출석 상태 Enum 작성

cookie-meringue committed:
test(red): 출석 상태 응답 레드 테스트 작성

cookie-meringue committed:
feat(green): 출석 상태 응답 기능 구현

cookie-meringue committed:
feat: 크루 상태 Enum 작성

cookie-meringue committed:
test(red): 출석 상태 리스트를 통해 초기화 레드 테스트 작성

cookie-meringue committed:
feat(green): 출석 상태 리스트를 통해 초기화 기능 구현

cookie-meringue committed:
docs: 오타 수정

cookie-meringue committed:
feat: 공휴일 Enum 작성

cookie-meringue committed:
test(red): 특정 일시가 공휴일인지 판단 기능 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 일시가 공휴일인지 판단 기능 구현

cookie-meringue committed:
docs: AttendanceLog 생성 방식 구현 사항 추가

cookie-meringue committed:
feat: 출석 기록 클래스 작성

cookie-meringue committed:
docs: 출석 기록 생성 매개변수 구현 사항 수정

cookie-meringue committed:
test(red): 출석 기록 생성 레드 테스트 작성

- LocalDateTime, CampusOperationPolicy 을 받아 생성된다.
- LocalDate 를 받아 결석 객체가 생성된다.

cookie-meringue committed:
feat(green): 출석 기록 생성 기능 구현

cookie-meringue committed:
refactor: 캠퍼스 운영 시간 반별 시 LocalDateTime 을 매개변수로 받도록 변경

cookie-meringue committed:
feat: AttendanceLog 일급 컬렉션 작성

cookie-meringue committed:
test(red): 특정 기간 사이 출석 기록 조회 레드 테스트 작성

cookie-meringue committed:
refactor: 캠퍼스 운영 판정 기능 함수 분리 및 가독성 개선

cookie-meringue committed:
refactor: 캠퍼스 운영 날짜 여부 판단 함수를 public 으로 변경

cookie-meringue committed:
feat: 같은 날짜인지 판단하는 함수 추가

cookie-meringue committed:
feat(green): 특정 기간 사이 출석 기록 조회 기능 구현

cookie-meringue committed:
docs: 출석 기록 통계 구현 사항 수정

cookie-meringue committed:
test(red): 출석 기록 통계 레드 테스트 작성

cookie-meringue committed:
feat: 리스트를 통해 출석 상태 통계 반환 함수 구현

cookie-meringue committed:
feat(green): 출석 기록 통계 기능 구현

cookie-meringue committed:
chore: 출석 기록 패키지 구조화

cookie-meringue committed:
feat: 크루 출석 기록 클래스 작성

cookie-meringue committed:
test(red): 크루 출석 기록 추가 레드 테스트 작성

cookie-meringue committed:
feat(green): 크루 출석 기록 추가 기능 구현

cookie-meringue committed:
docs: 출석일, 출석시 원시값 포장 구현 사항 추가 작성

cookie-meringue committed:
fix: 잘못된 진리 연산자 수정

- and 로 되어 있었어야 했음.

cookie-meringue committed:
docs: 출석 일시 원시값 포장 구현 사항 추가 작성

cookie-meringue committed:
feat: 출석 일자 클래스 작성

cookie-meringue committed:
test(red): 특정 일자가 출석 일자 vo 와 동일한 일자인지 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 일자가 출석 일자 vo 와 동일한 일자인지 판단 기능 구현

cookie-meringue committed:
test(red): 특정 DayOfWeek 가 출석 일자 vo 와 동일한 DayOfWeek 인지 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 DayOfWeek 가 출석 일자 vo 와 동일한 DayOfWeek 인지 판단 기능 구현

cookie-meringue committed:
style: AttendanceDateTest 내부 불필요한 주석 제거

cookie-meringue committed:
feat: 출석 시간 클래스 작성

cookie-meringue committed:
test(red): LocalTime 과 CampusOperationPolicy 를 받아 생성 레드 테스트 작성

cookie-meringue committed:
feat(green): LocalTime 과 CampusOperationPolicy 를 받아 생성 기능 구현

cookie-meringue committed:
docs: 값이 null 인 AttendanceTime 생성 구현 사항 수정

- CampusOperationPolicy 를 사용할 이유가 없다고 판단.

cookie-meringue committed:
test(red): 값이 null 인 AttendanceTime 생성 레드 테스트 작성

cookie-meringue committed:
feat(green): 값이 null 인 AttendanceTime 생성 기능 구현

cookie-meringue committed:
test(red): LocalTime 이 자신의 value 보다 이전인지 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): LocalTime 이 자신의 value 보다 이전인지 반환 기능 구현

cookie-meringue committed:
docs: 출석 시간 구현 사항 수정

- 자신의 value 가 비어 있는 null 한 시간인지 반환 동작 추가.

cookie-meringue committed:
test(red): 자신의 값이 null 인지 여부 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 자신의 값이 null 인지 여부 반환 기능 구현

cookie-meringue committed:
refactor: 캠퍼스 운영 시간 여부 반환 함수를 public 으로 변경

- 커밋 조정 중 누락된 커밋이었음.

cookie-meringue committed:
feat: 출석 일시 클래스 작성

cookie-meringue committed:
test(red): 일시와 CampusPolicy 를 통해 AttendanceDateTime 생성 레드 테스트 작성

cookie-meringue committed:
feat(green): 일시와 CampusPolicy 를 통해 AttendanceDateTime 생성 기능 구현

cookie-meringue committed:
test(red): 날짜와 CampusPolicy 를 통해 AttendanceDateTime 생성 레드 테스트 작성

cookie-meringue committed:
feat(green): 날짜와 CampusPolicy 를 통해 AttendanceDateTime 생성 기능 구현

cookie-meringue committed:
test(red): 특정 시간이 자신의 출석 시간보다 이전인지 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 시간이 자신의 출석 시간보다 이전인지 반환 기능 구현

cookie-meringue committed:
test(red): 특정 시간이 자신의 출석 일자인지 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 시간이 자신의 출석 일자인지 반환 기능 구현

cookie-meringue committed:
test(red): 자신의 출석 시간이 value 가 비어 있는 null 한 시간인지 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 자신의 출석 시간이 value 가 비어 있는 null 한 시간인지 반환 기능 구현

cookie-meringue committed:
test(red): 특정 요일이 요일이 자신의 요일인지 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 요일이 요일이 자신의 요일인지 반환 기능 구현

cookie-meringue committed:
refactor: 출석 일시 클래스를 사용하도록 변경

cookie-meringue committed:
test: private 메서드를 public 으로 변경함에 따른 단위 테스트 작성

cookie-meringue committed:
refactor: 출석 상태 통계 시, 출석 상태를 명시적으로 반환하도록 변경

- Enum 의 ordinal 에 의존하지 않도록 변경.

cookie-meringue committed:
style: 의미 없는 주석 제거

cookie-meringue committed:
chore: 출석 csv 파일 생성

cookie-meringue committed:
feat: 크루 출석 기록 역직렬화기 클래스 작성

cookie-meringue committed:
test(red): 크루 출석 기록 역직렬화 레드 테스트 작성

cookie-meringue committed:
feat(green): 크루 출석 기록 역직렬화 기능 구현

cookie-meringue committed:
docs: 크루 출석 기록 정렬 구현 사항 제거

- 보여주기 위한 정렬은 View 의 책임이라 판단.

cookie-meringue committed:
docs: 누락되어있던 출석 기록 추가 구현 사항 추가

cookie-meringue committed:
docs: 출석 기록 일급 컬렉션 동작 구현 사항 추가

cookie-meringue committed:
test(red): 지각, 결석 횟수 반환 레드 테스트 작성

cookie-meringue committed:
feat(green): 지각, 결석 횟수 반환 기능 구현

cookie-meringue committed:
feat: 지각, 결석 횟수 반환 기능 추가

cookie-meringue committed:
chore: 출석 상태 Enum 을 올바른 패키지로 이동

cookie-meringue committed:
docs: CrewAttendanceRepository 구현 사항 수정

cookie-meringue committed:
feat: 출석 기록 포함 여부 반환 함수 추가

- 복잡하지 않아서 테스트는 작성하지 않음.

cookie-meringue committed:
feat: 크루 출석 기록 저장소 클래스 작성

cookie-meringue committed:
test(red): 크루 출석 기록 저장소 레드 테스트 작성

cookie-meringue committed:
docs: AttendanceLogs 추가 구현 사항 추가

- 특정 AttendanceLog 의 포함 여부를 반환한다.

cookie-meringue committed:
test(red): 특정 AttendanceLog 수정 레드 테스트 작성

cookie-meringue committed:
feat(green): 특정 AttendanceLog 수정 기능 구현

cookie-meringue committed:
feat(green): CrewAttendanceRepository 기능 구현

cookie-meringue committed:
feat: Model-View-Controller 연결

cookie-meringue committed:
refactor: 함수명 오타 수정

cookie-meringue committed:
docs: 크루 출석 기록 저장소 구현 사항 수정

- CrewAttendanceLog 대신 Map<Crew, AttendanceLogs> 를 저장하도록 변경.

cookie-meringue committed:
refactor: 의미 없는 getter 제거

cookie-meringue committed:
refactor: 출석 상태 통계 반환 함수 내부 코드 가독성 향상

cookie-meringue committed:
refactor: record 로 변경

cookie-meringue committed:
refactor: CrewAttendanceRepository 가 내부 데이터를 List 대신 Map 으로 관리하도록 변경

cookie-meringue committed:
remove: 사용하지 않는 CrewAttendanceLog 제거

soominsohn commented on README.md:
```diff
@@ -43,3 +43,125 @@
 - 자세한 입력 예시는 실행 결과 예시를 참고하며, 프로그램을 시작하면 src/main/resources/attendances.csv를 통해 구현에 필요한 정보를 조회한다.
     - 닉네임과 출석 일시가 기록되어 있다.
+
+# Model 구현 클래스 목록
```
미리 정리를 해두셨군요 👍

soominsohn commented on src/main/java/attendance/model/campus/CampusOperationPolicy.java:
```diff
@@ -0,0 +1,36 @@
+    private static final LocalTime DEFAULT_OPEN_TIME = LocalTime.of(8, 0);
+    private static final LocalTime DEFAULT_CLOSE_TIME = LocalTime.of(23, 0);
+
+    public boolean isCampusOpen(final LocalDateTime dateTime) {
+        return isOpenDate(dateTime.toLocalDate()) && isOpenTime(dateTime.toLocalTime());
+    }
```
`isCampusOpen` 메서드가 테스트에서만 사용되고 서비스코드에서는 사용되지 않고 있어요.
코드 내용은 commit 기록에 다 남겨져있으니 더이상 사용되지 않는 코드는 과감하게 삭제해보면 어떨까요?

cookie-meringue replied:
정신없이 리팩터링하느랴 몰랐네요 ㅎㅎ 바로 삭제하겠습니다!

soominsohn commented on src/main/java/attendance/model/campus/CampusOperationPolicy.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+public class CampusOperationPolicy {
```
`CampusOperationPolicy`가 사용하는 곳에서 메서드마다 전달되고 있는 구조여서 코드가 복잡해 보일 수 있어요.
CampusOperationPolicy가 날짜 비교 역할만 수행한다면, 메서드를 static으로 만들어 유틸리티 클래스처럼 사용하는 방법도 고려해보면 좋을 것 같아요.
즉, CampusOperationPolicy가 객체 상태를 가질 필요가 없다면 인스턴스를 만들지 않고 static 메서드로 제공하는 것이 더 효율적일 수 있습니다.

cookie-meringue replied:
유틸리티 클래스도 좋은 선택 중 하나일 것 같네요 ㅎㅎ
그런데 유틸리티 클래스가 비즈니스 로직(캠퍼스 운영 일시와 관련된 로직)을 가지는 것은 어떻게 생각하시나요?
저는 유틸리티 클래스가 정말 최소 단위의 메서드를 제공해야 한다고 생각했었습니다.
ex) `isDateBefore(LocalDate a, LocalDate b)`

그런데, `CampusOperationPolicy` 가 유틸리티 클래스가 되면 캠퍼스 운영 시간이라는 비즈니스 로직을 가진 유틸리티 클래스가 탄생하고, 결과적으로 객체지향을 해치지 않을까? 하는 생각이 들었습니다.

물론, 현재 구조 자체에 문제가 있어.. 현 상황에서는 유틸리티 클래스 분리는 좋은 방법이라 생각합니다.

soominsohn replied:
오호.. 그렇군요 머랭말을 듣고나니 유틸리티 클래스가 비즈니스 로직을 가지고 있는 것도 어색하게 느껴지네요.
그렇다면 지금 구조도 괜찮다고 생각합니다!

soominsohn commented on src/main/java/attendance/model/attendance/log/AttendanceLog.java:
```diff
@@ -0,0 +1,88 @@
+    private final AttendanceDateTime attendanceDateTime;
+    private final AttendanceStatus attendanceStatus;
+
+    private AttendanceLog(AttendanceDateTime attendanceDateTime, AttendanceStatus attendanceStatus) {
+        this.attendanceDateTime = attendanceDateTime;
+        this.attendanceStatus = attendanceStatus;
+    }
```
정적팩토리 메서드를 사용해 기본생성자는 private으로 만들어주셨군요. 👍

cookie-meringue replied:
이 부분에 궁금증이 많았습니다!

정적 팩토리를 만들 때에 기본 생성자를 private 으로 하는 이유는 의도하지 않은 생성자를 외부에서 호출하는 것을 방지하는 것을 위해서라고 생각하는데요

하지만, record 를 사용할 때에 문제가 되었습니다. record 가 주는 편리한 기능 덕분에 record 도 사용하고자 했으나, 기본 생성자를 private 으로 만들지 못한다는 문법적 한계 때문에 record 를 사용하지 못한 경우가 많았습니다.

앤지의 생각은 어떠신가요?
1. 앤지가 생각하는 기본 생성자를 private 으로 만드는 이유는 무엇인가요?
2. record 사용 시 기본 생성자를 private 으로 할 수 없다는 제약 조건이 있는데, 앤지라면 다음 보기 중 어떤 것을 사용하실 건가요?
- 상황 : 정적 팩토리 메서드를 사용하고자 계획 중.
- 보기1 : record 를 사용하지 않고 class 를 사용해 기본 생성자를 private 으로 만들기.
- 보기2 : record 를 사용하고 기본 생성자도 public 으로 만들기.

soominsohn replied:
저라면 보기1번을 택하겠습니다.
record는 불변 객체를 간단하게 표현하는 데 매우 유용하지만, 생성자 통제가 어렵기 때문에 비즈니스 로직이 들어가는 도메인 객체보다는 필드를 전달하는 단순 데이터 구조(DTO)를 다룰 때 사용할 것 같아요.

cookie-meringue replied:
감사합니다!

soominsohn commented on src/test/java/attendance/model/attendance/repository/CrewAttendanceRepositoryTest.java:
```diff
@@ -0,0 +1,143 @@
+    void add() {
+
+        // Given
+        final Crew crew = new Crew("쿠키");
+        final AttendanceLog attendanceLog = AttendanceLog.fromDateTime(
+                LocalDateTime.of(2024, 12, 9, 10, 5),
+                campusOperationPolicy
+        );
```
자주 사용되는 테스트 메서드의 변수를 TestFixture로 따로 선언하면, 코드 작성량을 줄일 수 있어요!
그렇게 하면 반복적인 작업이 줄어들어, 테스트 작성이 조금 더 수월하고 덜 지루하게 느껴질 수도 있지 않을까요? 😊

cookie-meringue replied:
이번에 많이 느꼈습니다 ㅠㅠ 다음 미션부터는 Fixture 를 적극적으로 사용해보려고 합니다!

soominsohn commented on src/main/java/attendance/model/attendance/datetime/date/AttendanceDate.java:
```diff
@@ -0,0 +1,67 @@
+package attendance.model.attendance.datetime.date;
```
출석 시간을 관리하기 위해 `datetime`, `date`, `time` 패키지가 존재하고,
그 하위에 각각 `AttendanceDateTime`, `AttendanceDate`, `AttendanceTime` 객체가 있는 구조네요.
객체의 책임을 적절히 분리하는 것은 좋은 접근이지만,
상황에 따라 오히려 복잡도를 증가시키고 유지보수에 더 많은 시간이 소요될 수도 있어요.
구조를 유지하는 것이 정말 필요한지, 혹은 더 단순한 방식으로 관리할 수 있을지 고민해보는 것도 좋을 것 같습니다. 😊

cookie-meringue replied:
감사합니다! 더 생각해보겠습니다

cookie-meringue replied:
생각해보니 AttendanceDate, AttendanceTime 의 역할을 AttendanceDateTime 이 전부 책임져도 큰 문제가 없는 상황 같네요!
AttendanceDate, AttendanceTime 은 지워도 될 것 같습니다 ㅎㅎ

soominsohn commented on README.md:
```diff
@@ -1,3 +1,186 @@
 # java-attendance
 
 출석 미션 저장소
+
+# 기능 요구 사항
+
+- ## 출석 관리 규칙 및 시스템 설계 정책
+
```
주신 질문들에 대한 답변 드려요!

#### 2. TDD 적인가?
##### 2-1. 클래스를 세세하게 설계하여 오래걸리지 않았나?
세세한 부분 말고, 핵심적인 부분을 먼저 설계하고 작성해보면 어떨까요? 커밋로그를 보며 예를 들면, 크루 이름 길이 같은 경우는 핵심 기능이 아니니 나중에 분리하고 리팩토링 하면 좋을 것 같아요. 
##### 2-2. 이미 작성해놓은 테스트가 많으니 구조적 리팩터링을 적용하려면 리소스가 두 배로 들었습니다. 
코드를 작성하다 보면 설계가 잘못되었음을 발견하여 구조적인 리팩토링을 해야할 것 같을 때가 있습니다. 물론 처음 짠 코드 그대로 진행되면 좋겠지만, 현실적으로 수정하는 데 시간이 드는 것은 자연스러운 일이죠. 오히려 이러한 수정 과정도 코드 작성의 일부라고 생각해 보면 어떨까요? 원고를 작성할 때도 초고를 쓰는 것보다 수정하고 다듬는 데 더 많은 시간이 걸리듯, 코드 역시 처음부터 완벽할 수 없기에 리팩터링 과정에 시간이 더 드는 것이 당연할 수 있습니다.

#### 3. TDD로 개발하니 오히려 개발 시간이 너무 늘었습니다.
TDD를 적용하면서 개발 시간이 길어지고 재미가 없었다면, TDD 방식이 머랭에게 잘 맞지 않거나, 기존 방식이 더 익숙한 상태에서 갑자기 TDD를 시도하다 보니 적응이 어려웠을 수도 있어요. 그렇다면 TDD를 적용 방식을 조금 유연하게 조정해보는 것도 방법이에요.
처음부터 완벽한 테스트를 작성하려 하기보다는, 핵심적인 기능을 검증하는 최소한의 테스트부터 작성하고, 점진적으로 테스트 코드도 리팩터링해보는 건 어떨까요?
또 한 번 생각해 보면, TDD를 사용하지 않는다면, 작성한 기능이 정상적으로 동작하는지 즉각적으로 확인하기 어려울 수도 있어요. TDD에는 분명 단점도 있지만, 테스트가 빠른 피드백을 제공하고, 코드의 안정성을 높이는 장점도 있답니다. 
이번 미션을 통해 직접 체험하고 고민해 본 것만으로도 충분한 의미가 있다고 생각합니다. 이번 경험을 바탕으로 머랭에게 맞는 개발 방식을 찾아가면 될 것 같아요.

cookie-meringue replied:
감사합니다!

soominsohn reviewed (changes requested):
안녕하세요 머랭! 2단계 TDD로 진행 잘 해주셨어요.
주신 질문들에 대한 답변 남겨놨으니 확인해주시면 고맙겠습니다.
리팩토링 피드백이 필요하시면 진행해주시고 리뷰요청 주세요!

cookie-meringue committed:
remove: 사용하지 않는 isCampusOpen 메서드 제거

cookie-meringue committed:
refactor: 뷸필요한 클래스 분리를 하지 않도록 변경

- AttendanceDate, AttendanceTime 으로 분리하지 않음.
- 대신, AttendanceDateTime 이 전부 책임지도록 변경.

soominsohn commented on src/main/java/attendance/model/attendance/datetime/AttendanceDateTime.java:
```diff
@@ -0,0 +1,111 @@
+public class AttendanceDateTime {
+
+    private final LocalDate date;
+    private final LocalTime time;
```
time 필드가 null일 가능성이 있다면, 필드를 Optional<LocalTime>이나 default값으로 관리해보면 어떨까요? 
값이 없음을 표현할 때 null을 사용하면 어떤 위험을 가질 수 있을지에 대해 고민해보시면 좋을 것 같아요. 
`+` Optional도 주의해서 사용하시면 좋습니다. [참고](https://mangkyu.tistory.com/203)

cookie-meringue replied:
와우 이 부분은 전혀 생각하지 못했던 것 같습니다..
제가 Optional 을 설계에 맞지 않게 사용했네요
고쳐보도록 하겠습니다!

cookie-meringue replied:
공부할 거리를 주셔서 감사합니다😁

soominsohn commented on src/main/java/attendance/model/attendance/datetime/AttendanceDateTime.java:
```diff
@@ -0,0 +1,111 @@
+    public static AttendanceDateTime policyAppliedWithNullTime(
+            final LocalDate date,
+            final CampusOperationPolicy campusOperationPolicy
+    ) {
+
+        validateDate(date, campusOperationPolicy);
```
메서드 첫 번째 라인이 모두 개행으로 시작하는데 이유가 따로 있을까요?

cookie-meringue replied:
120라인을 넘어 매개변수를 개행 처리해서 받는 경우는 첫 번째 라인을 개행시키자는 저만의 컨벤션입니다!
가독성이 별로인가요? ㅎㅎ..
```
public a(int b){
// 개행 없이 코드 작성
}

public tooooooooooLongMethod(
    int toooooooLongParameter1,
    int toooooooLongParameter2
){

// 개행 후 코드 작성
}
```

soominsohn replied:
아하! 머랭의 컨벤션이었군요 ㅎㅎ!
컨벤션을 일관적으로 지키면서 코딩하신건 매우 좋지만, 일반적으로는 첫 번째 매개변수는 개행하지 않는 경우가 많아요. 😃
개인 컨벤션으로는 괜찮다고 생각하지만 팀프로젝트를 할 경우에는 IDE 자동 포맷팅을 사용하는 경우 충돌이 일어날 수 있기 때문에 협의하고 진행하면 좋을 것 같아요.

cookie-meringue replied:
감사합니다 ㅎㅎ 포매터를 작 적용해야겠네요

soominsohn commented on src/main/java/attendance/model/attendance/log/AttendanceLogs.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    private boolean isLate(AttendanceLog attendanceLog) {
+        return attendanceLog.getAttendanceStatus() == AttendanceStatus.LATE;
```
```suggestion
        return attendanceLog.isLate();
```
AttendanceLog에 AttendanceStatus 비교 메서드를 만들면 AttendanceLogs에 isLate 메서드를 만들지 않아도 stream에서 처리해볼 수 있을 것 같아요. 😄 
아래 isAbsence 메서드도 마찬가지입니다!

cookie-meringue replied:
감사합니다 ㅎㅎ 설계 중에 놓친 부분이 아직 꽤 있네요...

soominsohn replied:
저도 맨날 놓치고 나중에 리팩토링 한답니다 ㅋㅋ.. 👍

soominsohn commented on src/main/java/attendance/model/attendance/log/AttendanceLogs.java:
```diff
@@ -0,0 +1,151 @@
+        return attendanceLog.getAttendanceStatus() == AttendanceStatus.ABSENCE;
+    }
+
+    public void update(final AttendanceLog from, final AttendanceLog to) {
+        final Optional<AttendanceLog> targetAttendanceLog = values.stream()
+                .filter(log -> log.isSameDate(from.getDate()))
+                .findFirst();
+
+        if (targetAttendanceLog.isPresent()) {
+            values.remove(targetAttendanceLog.get());
+            values.add(to);
+            return;
+        }
+
+        throw new IllegalArgumentException("해당 날짜의 출석 기록이 존재하지 않습니다.");
+    }
```
```suggestion
    public void update(final AttendanceLog from, final AttendanceLog to) {
        final AttendanceLog targetAttendanceLog = values.stream()
                .filter(log -> log.isSameDate(from.getDate()))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("해당 날짜의 출석 기록이 존재하지 않습니다."));
        
        values.remove(targetAttendanceLog);
        values.add(to);
    }
```
targetAttendanceLog를 Optional로 반환받지 않아도 로직상 문제가 없을 것 같은데 어떤가요? 😄

cookie-meringue replied:
맞습니다 ㅎㅎ 성숙하지 못한 코드네요

soominsohn replied:
아닙니당 이런건 혼자 보면 잘 안보이더라구요

soominsohn commented on src/main/java/attendance/model/crew/CrewStatus.java:
```diff
@@ -0,0 +1,53 @@
+                attendanceStatuses.stream()
+                        .filter(AttendanceStatus.LATE::equals)
+                        .count()
+        ) / 3;
```
지각 3회 == 1결석을 상수화하는 것도 좋을 것 같아요.`LATE_TO_ABSENCE_RATIO` 매직넘버 추출 기준을 생각할 때 변경가능성과 다른사람이 읽었을 때 설명없이 이해할 수 있는가를 고려하면 좋습니다.

cookie-meringue replied:
네 ㅎㅎ 감사합니다.
설계가 잘못되었던 부분이라 리팩터링했습니다!
fdbfb17e4547ca4265313ae738aa8919441f0897
0260e6629d3ced22826852c660c377ae1f99d3ab

soominsohn commented on src/main/java/attendance/model/attendance/repository/CrewAttendanceRepository.java:
```diff
@@ -0,0 +1,77 @@
+        findByCrew(crew).update(from, to);
+    }
+
+    public AttendanceLogs findByCrewFromTo(
```
FromTo도 명확한 표현이긴 하지만 코드상에서 날짜 사이를 표현할 때는 between을 더 많이 사용합니다. between을사용해보는건 어떨까요?

cookie-meringue replied:
아..! between
제가 딱 원하던 단어네요 수정하겠습니다 ㅎㅎ

soominsohn commented on src/main/java/attendance/model/attendance/repository/CrewAttendanceRepository.java:
```diff
@@ -0,0 +1,77 @@
+
+public class CrewAttendanceRepository {
+
+    private final Map<Crew, AttendanceLogs> crewAttendanceLogs;
```
Map을 사용해 효율적으로 관리할 수 있겠네요 👍

soominsohn reviewed (changes requested):
안녕하세요 머랭~
리뷰반영 빠르게 해주셨네요. 고맙습니다.
질문에 대한 답변과 간단한 코멘트 몇개 남겼으니 확인부탁드려요!
마지막까지 파이팅입니다 👍

cookie-meringue committed:
feat: 지각, 결석 판별 함수 추가

cookie-meringue committed:
refactor: AttendanceLog 의 지각, 결석 판정 함수를 이용하도록 변경

cookie-meringue committed:
refactor: 의미 없는 Optional 을 사용하지 않도록 변경

cookie-meringue committed:
refactor: 특정 일 사이 값에 대한 함수명을 between 으로 변경

- FromTo 보다 더 좋은 의미 전달을 위함.

cookie-meringue committed:
refactor: AttendanceLogs 를 통해 CrewStatus 를 얻도록 변경

cookie-meringue committed:
refactor: 매직넘버 상수화

soominsohn reviewed (approved):
머랭~! 2단계 리팩토링까지 정말 잘해주셨어요.
머랭이 질문 많이 해주신 덕분에 저도 많이 배웠답니다.
다음 미션도 열심히 재밌게 코딩하길 바랍니다~!
파이팅하세요~.
